In [1]:
import os
import random
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
import matplotlib.pyplot as plt
import numpy as np
from tensorflow.keras.datasets import cifar10
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.preprocessing.image import ImageDataGenerator

In [2]:
(X_train, Y_train),(X_test, Y_test) = cifar10.load_data()
print(X_train.shape)
print(X_test.shape)



(50000, 32, 32, 3)
(10000, 32, 32, 3)


In [3]:
gpu_device = tf.config.experimental.list_physical_devices('GPU')
print(f"Number of GPU = {len(gpu_device)}")
tf.config.experimental.set_memory_growth(gpu_device[0], True)

Number of GPU = 1


In [4]:
X_valid, Y_valid = X_train[:5000], Y_train[:5000]

In [5]:
X_train = X_train / 255.0
X_test = X_test/ 255.0

In [6]:
model = keras.Sequential([
    keras.Input(shape=(32,32,3)),
    layers.Conv2D(filters=64, kernel_size=(3,3), padding="same", activation="relu"),
    layers.Conv2D(filters=64, kernel_size=(3,3), padding="same", activation="relu"),
    layers.MaxPooling2D(pool_size=(2,2), strides=(2,2)),
    
    layers.Conv2D(filters=128, kernel_size=(3,3), padding="same", activation="relu"),
    layers.Conv2D(filters=128, kernel_size=(3,3), padding="same", activation="relu"),
    layers.MaxPooling2D(pool_size=(2,2), strides=(2,2)),
    
    layers.Conv2D(filters=256, kernel_size=(3,3), padding="same", activation="relu"),
    layers.Conv2D(filters=256, kernel_size=(3,3), padding="same", activation="relu"),
    layers.Conv2D(filters=256, kernel_size=(3,3), padding="same", activation="relu"),
    layers.MaxPooling2D(pool_size=(2,2), strides=(2,2)),
    
    layers.Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"),
    layers.Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"),
    layers.Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"),
    layers.MaxPooling2D(pool_size=(2,2), strides=(2,2)),
    
    layers.Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"),
    layers.Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"),
    layers.Conv2D(filters=512, kernel_size=(3,3), padding="same", activation="relu"),
    layers.MaxPooling2D(pool_size=(2,2), strides=(2,2)),
    
    layers.Flatten(),
    layers.Dense(4096, activation='relu'),
    layers.Dense(4096, activation='relu'),
    layers.Dense(29, activation='softmax')
])
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 32, 32, 64)        1792      
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 32, 32, 64)        36928     
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 16, 16, 64)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 16, 16, 128)       73856     
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 16, 16, 128)       147584    
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 8, 8, 128)         0         
_________________________________________________________________
conv2d_4 (Conv2D)            (None, 8, 8, 256)         2

In [7]:
learning_rate = 0.0001
adam = Adam(lr=learning_rate)
model.compile(optimizer=adam, loss='sparse_categorical_crossentropy', metrics=['accuracy'])


In [8]:
h = model.fit(X_train, Y_train, epochs=6, validation_data=(X_valid, Y_valid), batch_size=128)

Epoch 1/6
391/391 [==============================] - 23s 49ms/step - loss: 2.3933 - accuracy: 0.1147 - val_loss: 147.0635 - val_accuracy: 0.2192
Epoch 2/6
391/391 [==============================] - 17s 43ms/step - loss: 1.6840 - accuracy: 0.3357 - val_loss: 86.6880 - val_accuracy: 0.3520
Epoch 3/6
391/391 [==============================] - 17s 44ms/step - loss: 1.3563 - accuracy: 0.4923 - val_loss: 76.3236 - val_accuracy: 0.4150
Epoch 4/6
391/391 [==============================] - 17s 44ms/step - loss: 1.1178 - accuracy: 0.5929 - val_loss: 60.8297 - val_accuracy: 0.4818
Epoch 5/6
391/391 [==============================] - 17s 44ms/step - loss: 0.9399 - accuracy: 0.6594 - val_loss: 77.6516 - val_accuracy: 0.4466
Epoch 6/6
391/391 [==============================] - 17s 44ms/step - loss: 0.8026 - accuracy: 0.7142 - val_loss: 51.4509 - val_accuracy: 0.5092


In [9]:
train_gen = ImageDataGenerator(
    featurewise_center=False,
    rotation_range=50,
    width_shift_range=0.2,
    horizontal_flip= True,
    fill_mode='nearest',
    zoom_range=0.1,
)
    

In [10]:
aug_train = train_gen.flow(X_train, Y_train, batch_size=128)

In [11]:
h = model.fit(aug_train, epochs=6, validation_data=(X_valid, Y_valid))

Epoch 1/6
391/391 [==============================] - 20s 50ms/step - loss: 1.2077 - accuracy: 0.5741 - val_loss: 88.8838 - val_accuracy: 0.4088
Epoch 2/6
391/391 [==============================] - 19s 50ms/step - loss: 1.1017 - accuracy: 0.6112 - val_loss: 98.9465 - val_accuracy: 0.4038
Epoch 3/6
391/391 [==============================] - 19s 50ms/step - loss: 1.0500 - accuracy: 0.6310 - val_loss: 84.9435 - val_accuracy: 0.4440
Epoch 4/6
391/391 [==============================] - 19s 50ms/step - loss: 0.9894 - accuracy: 0.6542 - val_loss: 103.6626 - val_accuracy: 0.4472
Epoch 5/6
391/391 [==============================] - 20s 51ms/step - loss: 0.9630 - accuracy: 0.6641 - val_loss: 121.1412 - val_accuracy: 0.4086
Epoch 6/6
391/391 [==============================] - 21s 52ms/step - loss: 0.9214 - accuracy: 0.6805 - val_loss: 124.1379 - val_accuracy: 0.3918


Difference:


We see some major difference in both training with data augmentation and without augmentation. In without augmentation we see training accuracy increase linearly and reched 71% at 6 epochs also we see same thing in validation accuracy. When we do augmentation in data it also linearly increase accuracy in training data but it reached 92% at 6 epochs where without augmentation it reached 71% so, here a major difference is made in training data. In validation accuracy we see some difference are also made in with augmentation accuracy are very high at the first epochs as time go accuracy was dropped and reached 39% accuracy where we see a linear increasing accuracy without augmentation and reached 50% accuracy. Those are the major difference we see in with and without augmentation. 